validator/sawtooth_validator/journal/publisher.py

```diff
@@ -744,12 +742,30 @@ def has_batch(self, batch_id):
         with self._lock:
             if batch_id in self._pending_batches:
                 return True
-            if batch_id in self._queued_batch_ids:
+            if batch_id in self._batch_queue:
                 return True
 
         return False
 
 
+class IncomingBatchQueue:
+    def __init__(self):
+        self._queue = queue.Queue()
+        self._ids = ConcurrentSet()
+
+    def put(self, batch):
+        self._ids.add(batch.header_signature)
```

Comment:
If these needs to be kept in sync, atomically, and this and `get` are used across threads, there needs to be a lock around the actions here.  Based on your commit message this doesn't guarantee they stay in sync, from a thread perspective.